api/envoy/config/filter/udp/dns_filter/v2alpha/dns_filter.proto

```diff
@@ -0,0 +1,64 @@
+syntax = "proto3";
+
+package envoy.config.filter.udp.dns_filter.v2alpha;
+
+import "google/protobuf/duration.proto";
+
+import "udpa/annotations/status.proto";
+
+import "udpa/annotations/migrate.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.config.filter.udp.dns_filter.v2alpha";
+option java_outer_classname = "DnsFilterProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_migrate).move_to_package =
+    "envoy.extensions.filters.udp.dns_filter.v3";
+option (udpa.annotations.file_status).work_in_progress = true;
+
+// Configuration for the DNS filter.
+message DnsFilterConfig {
+  // Specify the method of determining the address of a given endpoint
+  enum EndpointType { STATIC = 0; }
+
+  // This message type is extensible and can contain a list of addresses
+  // or dictate some other method for resolving the addresses for an
+  // endpoint
+  message DnsEndpoint {
+    EndpointType type = 1;
+
+    // A list of IP addresses Envoy returns to queries for the given domain
+    repeated string address = 2;
+  }
+
+  message DnsVirtualDomain {
+    // The domain name for which Envoy will respond to query requests
+    string name = 1 [(validate.rules).string = {min_len: 2}];
+
+    // The configuration containing the method to determine the address
+    // of this endpoint
+    DnsEndpoint endpoint = 2;
+  }
+
+  // This message contains the configuration for the Dns Filter operating
+  // in a server context. This message will contain the virtual hosts and
+  // associated addresses with which Envoy will respond to queries
+  message ServerCtxConfig {
+    uint32 retry_count = 1;
+
+    // Fully qualified domain names for which Envoy will respond to queries
+    repeated DnsVirtualDomain virtual_domains = 2 [(validate.rules).repeated = {min_items: 1}];
+
+    // Sets the TTL in dns answers from Envoy returned to the client
+    google.protobuf.Duration answer_ttl = 3 [(validate.rules).duration = {gt {}}];
+
+    // Seed the virtual domain configuration from an external path
```

Comment:
What does this do? Should this be a `DataSource`?

Reply:
It looks that way.  I can make that change.

Reply:
To answer your initial question, we wanted a way for the domain/address map to be defined and populated separately from the control plane.   The intent here is for us to define a file with this configuration and Envoy will load and use it instead of any mappings sent by the control plane.

Reply:
You probably want to use `oneof` to make it clear that it's either/or.

Reply:
One other thing to consider is the format of this file. I think it would be ideal to give it a proto format, living under the `envoy.data...` package namespace. That way, you can have JSON/YAML/proto representations for free.

If you do that, you will want to make the `oneof` alternative here a field with this `envoy.data...` type, for economy of data types.

In fact, since this is a `DataSource`, and `DataSource` supports inlining already, maybe there is no need at all to have a `oneof` or any inline config here, it should all come via a proto/JSON/YAML in the `DataSource`.

Reply:
@htuch, I agree about the file format.   I did update this to be a `DataSource`.   I'll update with `oneof` as well.